File: tests/pow-drivers/driver.py
#!/usr/bin/env python3
"""PoW driver for py/shavar.py. See tests/pow.sh.

Reads the vector file named on the command line and writes one
`id <TAB> met|unmet|invalid` line per vector to stdout. It is a marshalling
shim only: every decision comes from py/shavar.py.
"""
import os
import sys

sys.path.insert(0, os.path.join(
    os.path.dirname(os.path.abspath(__file__)), "..", "..", "py"))
import shavar  # noqa: E402


def main(argv):
    if len(argv) != 2:
        sys.stderr.write("usage: driver.py VECTORS.tsv\n")
        return 2
    with open(argv[1]) as f:
        for line in f:
            if line.startswith("#") or not line.strip():
                continue
            vid, dhex, nbhex = line.rstrip("\n").split("\t")[:3]
            digest = bytes.fromhex(dhex)
            nbits = int(nbhex, 16)
            try:
                verdict = "met" if shavar.pow_check(digest, nbits) else "unmet"
            except shavar.ShavarError:
                verdict = "invalid"
            sys.stdout.write("%s\t%s\n" % (vid, verdict))
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: tests/pow-drivers/driver.pl
#!/usr/bin/perl
# PoW driver for pl/shavar.pl. See tests/pow.sh.
#
# Reads the vector file named on the command line and writes one
# `id <TAB> met|unmet|invalid` line per vector. Marshalling only: every
# decision comes from pl/shavar.pl, pulled in through its modulino guard.
use strict;
use warnings;
use File::Basename qw(dirname);

my $here = dirname(__FILE__);
require "$here/../../pl/shavar.pl";

die "usage: driver.pl VECTORS.tsv\n" unless @ARGV == 1;
open my $fh, '<', $ARGV[0] or die "$ARGV[0]: $!\n";
while (my $line = <$fh>) {
    next if $line =~ /^#/ || $line =~ /^\s*$/;
    chomp $line;
    my ($id, $dhex, $nbhex) = (split /\t/, $line)[0, 1, 2];
    my $digest = pack('H*', $dhex);
    my $nbits  = hex($nbhex);
    my $verdict = eval { main::pow_check($digest, $nbits) ? 'met' : 'unmet' };
    $verdict = 'invalid' unless defined $verdict;
    print "$id\t$verdict\n";
}
close $fh;
